### PR TITLE
[RDY] Stop NPCs from stealing from your vehicles

### PR DESCRIPTION
**Pull request by angoddu**
I have added the ability to lock all cargo parts through the interact menu. Each cargo part can be individually locked. Intentionally made it part specific to make it harder to abuse the system. 

This has been a long standing bug and I decided that it was worth trying to tackle. NPCs will no longer steal from your shopping cart so long as it is locked.

Let me know if anything should be change.

### CODE REVIEW

**Comment by Coolthulhu** [comment]
"Locking" a shopping cart?

Let's just drop the NPC pickup from vehicles if it's such a problem.
Having to manually lock every cargo tile sounds terrible.

Alternatively, allow "claiming" a remembered vehicle. A number of claimed vehicles could be limited per overmap (remembered vehicles are visible from overmap level at the code side).
Same behavior could be made for camps: random NPCs would not touch anything on the map tile with a billboard and camp designated around it.

**Comment by angoddu** [comment]
It would be rather easy to stop NPCs from looting cars. It would be a simple solution but may be too broad. Would it mess up any kind of quest or something?

**Comment by kevingranade** [comment]
This is too broad, if you're going to lock vehicle tiles, there needs to be
a mechanism to do so such as a locked trunk or door, or with a shopping
cart a lid.

In other words, add a LOCKING_CARGO flag to appropriate vehicle parts and
use that to disallow NPC access.

From a UI point of view, just disallow NPC access if the flag is present,
no need to manually lock and unlock.

**Comment by angoddu** [comment]
I was thinking that the COVERED flag could also make the cargo inaccessible to NPCs. If it is able to block all light then it can make it so NPCs can't see or take these items

**Comment by kevingranade** [comment]
No... NPCs are people and they're just as capable of opening containers as the player is, COVERED doesn't cut it.

**Comment by kevingranade** [comment]
Build is failing, you didn't clean up all the manual lock/unlock changes and that's breaking compilation.
In general that code needs to be removed anyway.
The number of vehicle parts covered by this is still too high, stowboards, boxes, and cargo spaces don't necessarally have locks. I can see a box with a lock, but it needs to be a new vehicle part, not the default.  Stowboards and cargo spaces are just shelves and racks, they can't lock.

Also this needs to be surfaced to the player somehow, that can stay as a TODO for the moment since some players clearly expect NPCs to be unable to take things anyway, but we need to communicate to the player that some cargo locations are safe from NPCs and others aren't.

**Comment by angoddu** [comment]
Where would I add the TODO comment?

**Comment by angoddu** [comment]
npcmove.cpp is on the astyle-blacklist so do I need to worry that it failed it's astyle check? I only changed a few lines. Also I have no idea why this monster-test.cpp failed. I didn't touch it

**Comment by kevingranade** [comment]
Npcmove.cpp is no longer on the a style blacklist in the main branch.
Merge from or rebase on master, then run ```make astyle```.
The monster test failure is an intermittent upstream failure and isn't your problem.

File: data/json/vehicle_parts.json [hunk 1]
```diff
@@ -726,6 +726,22 @@
...
+        "broken_color" : "brown",
+        "damage_modifier" : 60,
+        "durability" : 100,
+        "size" : 275,
+        "item" : "frame",
+        "difficulty" : 2,
+        "location" : "center",
+        "flags" : ["CARGO", "CARGO_LOCKING", "TOOL_SCREWDRIVER", "BOARDABLE", "COVERED"],
+        "breaks_into": "ig_vp_frame"
+    },{
```
Comment (Coolthulhu) [inline data/json/vehicle_parts.json]: Vehicle parts support "copy-from". Don't specify everything from scratch.
Check out `data/json/vehicleparts/battery.json` for an example.

Comment (angoddu) [reply data/json/vehicle_parts.json]: Good point. I will do this

**Comment by angoddu** [comment]
Could someone re-run the travis test please?

**Comment by kevingranade** [comment]
Realized something else about this, we're having a lock (and key, but I'm less concerned about that) appear out of nowhere when you craft a frame into a trunk or hatch.  Hinges and a latch were already pushing it a bit, but now it's starting to get a bit silly.

In the end, this is mostly a bookkeeping issue since most frames are harvested from vehicles anyway, so presumably we're getting a stock of hinges, latches, handles, and locks from tearing down other vehicles, so we can call it "abstracted" for now, but this is something to track in the future that can expand vehicle crafting.

**Comment by angoddu** [comment]
I like this idea. It would be an interesting addition to vehicles

**Comment by Coolthulhu** [comment]
Now that I think about it, having to detach and reattach every vehicle part just to make it lockable sounds very tedious.
Would be better if those parts just had a "lock-installable" flag of some sort and there was a "lock" part.

**Comment by tinukedaya** [comment]
>Now that I think about it, having to detach and reattach every vehicle part just to make it lockable sounds very tedious.
>Would be better if those parts just had a "lock-installable" flag of some sort and there was a "lock" part.

I like this. It would also solve the "worry" @kevingranade had about abstracting too much.

**Comment by angoddu** [comment]
I could make a vehicle lock part and make the parts with the "CARGO_LOCKING" flag have a "LOCKABLE_CARGO" flag.

**Comment by angoddu** [comment]
I'm having an issue with my location flag. I have it coded like
`"location" : "on_lockable_cargo",`
but it is being installable anywhere. What am I missing with the locations? I thought that you could just have the location be the same as the flag with "on_" before it.

**Comment by angoddu** [comment]
I fixed the location error. I needed to add a condition to the vehicle::can_mount() method since location isn't flexible like I thought.

**Comment by cainiaowu** [comment]
Jenkins, rebuild.

**Pull request merged**